File: dashboard/backend/evaluationplane/metric_analysis_catalog_test.go
package evaluationplane

import (
	"bytes"
	"crypto/sha256"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"slices"
	"sort"
	"testing"
)

func TestMetricAnalysisCatalogPackagedMirrorsAreByteIdentical(t *testing.T) {
	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("resolve metric analysis catalog test location")
	}
	directory := filepath.Dir(currentFile)
	paths := []string{
		filepath.Join(directory, "../../../src/vllm-sr/cli/evaluation/golden/metric_analysis_catalog.v1.json"),
		filepath.Join(directory, "../../frontend/src/contracts/metric_analysis_catalog.v1.json"),
	}
	for _, path := range paths {
		mirrored, err := os.ReadFile(path)
		if err != nil {
			t.Fatalf("read metric analysis catalog mirror %q: %v", path, err)
		}
		if !bytes.Equal(metricAnalysisCatalogJSON, mirrored) {
			t.Fatalf("metric analysis catalog mirror %q differs from the embedded bytes", path)
		}
	}
	want := "35ec19955f0ac793cabb9ae986baa96f5403284790784a0c64a3b6af4a4a533a"
	if got := fmt.Sprintf("%x", sha256.Sum256(metricAnalysisCatalogJSON)); got != want {
		t.Fatalf("metric analysis catalog digest=%s, want %s", got, want)
	}
}

func TestMetricAnalysisCatalogResolvesExactAndTypedDynamicMetrics(t *testing.T) {
	if err := ValidateMetricAnalysisCatalog(); err != nil {
		t.Fatalf("metric analysis catalog is invalid: %v", err)
	}
	if len(metricAnalysisCatalogData.document.StaticMetrics) != 132 || len(metricAnalysisCatalogData.document.DynamicFamilies) != 6 {
		t.Fatalf("metric analysis universe has unexpected cardinality")
	}
	staticIDs := make([]string, 0, len(metricAnalysisCatalogData.document.StaticMetrics))
	for _, metric := range metricAnalysisCatalogData.document.StaticMetrics {
		staticIDs = append(staticIDs, metric.ID)
	}
	if !sort.StringsAreSorted(staticIDs) {
		t.Fatal("metric analysis static ids are not sorted")
	}
	for _, metricID := range staticIDs {
		match, err := ResolveMetricAnalysisCatalog(metricID)
		if err != nil || match.MetricID != metricID || match.FamilyID != "" {
			t.Fatalf("resolve exact metric %q match=%+v err=%v", metricID, match, err)
		}
	}

	tests := []struct {
		id, family, analysisRef, analysisUnit, weighting string
	}{
		{"routing.accuracy", "", "routing.case.ratio", "route_case", "uniform_case"},
		{"model_pool.arm.fast.quality", "model-pool-arm", "model-pool.arm.case-mean", "pool_case", "uniform_case"},
		{"model_pool.arm.u-bW9kZWwudjE.quality", "model-pool-arm", "model-pool.arm.case-mean", "pool_case", "uniform_case"},
		{"capacity.level.16.latency_p95_ms", "capacity-level", "capacity.level.request-quantile", "measurement_request", "uniform_request"},
		{"routing_recipe.e1.signal.u-ZG9tYWluOnJlYXNvbmluZw.present_rate", "routing-recipe-e1-input", "routing-recipe.e1.input-rate", "decision_input", "uniform_case"},
		{"routing_recipe.e2.feasible_oracle_recall_at_64", "routing-recipe-e2-recall", "routing-recipe.e2.feasible-recall", "decision_case", "uniform_case"},
	}
	for _, test := range tests {
		match, err := ResolveMetricAnalysisCatalog(test.id)
		if err != nil || match.FamilyID != test.family || match.Specification.AnalysisRef != test.analysisRef ||
			match.Specification.AnalysisUnit != test.analysisUnit || match.Specification.Weighting != test.weighting {
			t.Fatalf("resolve %q match=%+v err=%v", test.id, match, err)
		}
	}

	for _, unknown := range []string{
		"routing.injected",
		"model_pool.arm.model.v1.quality",
		"model_pool.arm.u-abc.quality",
		"capacity.level.0.success_rate",
		"capacity.level.16.injected",
		"routing_recipe.e2.feasible_oracle_recall_at_65",
	} {
		if _, err := ResolveMetricAnalysisCatalog(unknown); !errors.Is(err, ErrUnknownMetricAnalysisID) {
			t.Fatalf("unknown metric %q error=%v", unknown, err)
		}
	}
}

func TestMetricAnalysisSubjectCodecGoldenVectors(t *testing.T) {
	if err := ValidateMetricAnalysisCatalog(); err != nil {
		t.Fatal(err)
	}
	for _, vector := range metricAnalysisCatalogData.document.IdentifierEncoding.Vectors {
		encoded, err := EncodeMetricAnalysisSubjectID(vector.Raw)
		if err != nil || encoded != vector.Encoded {
			t.Fatalf("encode %q=%q err=%v, want %q", vector.Raw, encoded, err, vector.Encoded)
		}
		decoded, err := DecodeMetricAnalysisSubjectID(vector.Encoded)
		if err != nil || decoded != vector.Raw {
			t.Fatalf("decode %q=%q err=%v, want %q", vector.Encoded, decoded, err, vector.Raw)
		}
	}
	for raw, want := range map[string]string{
		"domain:reasoning":      "u-ZG9tYWluOnJlYXNvbmluZw",
		"classifier:risk:RISKY": "u-Y2xhc3NpZmllcjpyaXNrOlJJU0tZ",
	} {
		got, err := EncodeMetricAnalysisSubjectID(raw)
		if err != nil || got != want {
			t.Fatalf("router key codec %q=%q err=%v, want %q", raw, got, err, want)
		}
	}
}

func TestMetricAnalysisResolverRejectsAmbiguity(t *testing.T) {
	if err := ValidateMetricAnalysisCatalog(); err != nil {
		t.Fatal(err)
	}
	copyIndex := metricAnalysisCatalogData
	copyIndex.document.DynamicFamilies = append(
		append([]metricAnalysisDynamicFamily(nil), metricAnalysisCatalogData.document.DynamicFamilies...),
		metricAnalysisCatalogData.document.DynamicFamilies[0],
	)
	if _, err := resolveMetricAnalysisCatalog(&copyIndex, "capacity.level.16.success_rate"); !errors.Is(err, ErrAmbiguousMetricAnalysisID) {
		t.Fatalf("ambiguous resolver error=%v", err)
	}
}

func TestMetricAnalysisRetiredDisplayIDsNeverResolve(t *testing.T) {
	if err := ValidateMetricAnalysisCatalog(); err != nil {
		t.Fatal(err)
	}
	if len(metricAnalysisCatalogData.document.RetiredMetricIDs) != 11 {
		t.Fatalf("retired metric count=%d", len(metricAnalysisCatalogData.document.RetiredMetricIDs))
	}
	for _, retired := range metricAnalysisCatalogData.document.RetiredMetricIDs {
		if _, err := ResolveMetricAnalysisCatalog(retired.ID); !errors.Is(err, ErrUnknownMetricAnalysisID) {
			t.Fatalf("retired metric %q resolved: %v", retired.ID, err)
		}
		if retired.Replacement != nil {
			if _, err := ResolveMetricAnalysisCatalog(*retired.Replacement); err != nil {
				t.Fatalf("retired replacement %q is not canonical: %v", *retired.Replacement, err)
			}
		}
	}
}

func TestBuiltinTrackCapabilitiesComeOnlyFromCanonicalMetricCatalog(t *testing.T) {
	if err := ValidateMetricAnalysisCatalog(); err != nil {
		t.Fatal(err)
	}
	seen := make(map[string]struct{}, len(metricAnalysisCatalogData.document.StaticMetrics))
	for _, track := range builtinTracks() {
		want := StaticMetricAnalysisIDsForTrack(track.ID)
		if !slices.Equal(track.Metrics, want) {
			t.Fatalf("track %q metrics=%v, want canonical %v", track.ID, track.Metrics, want)
		}
		for _, metricID := range track.Metrics {
			if _, duplicate := seen[metricID]; duplicate {
				t.Fatalf("canonical metric %q is advertised by multiple tracks", metricID)
			}
			seen[metricID] = struct{}{}
		}
	}
	if len(seen) != len(metricAnalysisCatalogData.document.StaticMetrics) {
		t.Fatalf("advertised %d canonical metrics, want %d", len(seen), len(metricAnalysisCatalogData.document.StaticMetrics))
	}
}

File: dashboard/frontend/src/contracts/metricAnalysisCatalog.test.ts
import { readFileSync } from 'node:fs'

import { describe, expect, it } from 'vitest'

import {
  decodeMetricAnalysisSubjectID,
  DYNAMIC_METRIC_ANALYSIS_FAMILY_IDS,
  encodeMetricAnalysisSubjectID,
  METRIC_ANALYSIS_CATALOG_SOURCE,
  resolveMetricAnalysisCatalog,
  RETIRED_METRIC_ANALYSIS_IDS,
  STATIC_METRIC_ANALYSIS_IDS,
  validateMetricAnalysisCatalogSource,
} from './metricAnalysisCatalog'

interface CatalogFixture {
  identifier_encoding: { vectors: Array<{ raw: string; encoded: string }> }
  dynamic_families: Array<{
    id: string
    literal_prefix: string
    pattern: string
    examples: Array<{
      metric_id: string
      captures: Record<string, string>
      analysis_ref: string
    }>
  }>
}

const catalog = JSON.parse(METRIC_ANALYSIS_CATALOG_SOURCE) as CatalogFixture

describe('canonical metric analysis catalog', () => {
  it('packages byte-identical Python, Go, and TypeScript resources', () => {
    const typescriptResource = readFileSync(
      new URL('./metric_analysis_catalog.v1.json', import.meta.url),
      'utf8',
    )
    const pythonResource = readFileSync(
      new URL(
        '../../../../src/vllm-sr/cli/evaluation/golden/metric_analysis_catalog.v1.json',
        import.meta.url,
      ),
      'utf8',
    )
    const goResource = readFileSync(
      new URL('../../../backend/evaluationplane/metric_analysis_catalog.v1.json', import.meta.url),
      'utf8',
    )

    expect(typescriptResource).toBe(pythonResource)
    expect(typescriptResource).toBe(goResource)
    expect(JSON.parse(METRIC_ANALYSIS_CATALOG_SOURCE)).toEqual(JSON.parse(typescriptResource))
  })

  it('resolves all 132 exact ids and all six typed dynamic families', () => {
    expect(STATIC_METRIC_ANALYSIS_IDS).toHaveLength(132)
    expect(DYNAMIC_METRIC_ANALYSIS_FAMILY_IDS).toHaveLength(6)
    expect(STATIC_METRIC_ANALYSIS_IDS).toEqual([...STATIC_METRIC_ANALYSIS_IDS].sort())
    expect(DYNAMIC_METRIC_ANALYSIS_FAMILY_IDS).toEqual(
      [...DYNAMIC_METRIC_ANALYSIS_FAMILY_IDS].sort(),
    )

    for (const metricID of STATIC_METRIC_ANALYSIS_IDS) {
      const match = resolveMetricAnalysisCatalog(metricID)
      expect(match.metric_id).toBe(metricID)
      expect(match.family_id).toBeUndefined()
    }
    for (const family of catalog.dynamic_families) {
      for (const example of family.examples) {
        const match = resolveMetricAnalysisCatalog(example.metric_id)
        expect(match.family_id).toBe(family.id)
        expect(match.captures).toEqual(example.captures)
        expect(match.specification.id).toBe(example.analysis_ref)
      }
    }
  })

  it('uses a canonical one-segment codec, including Router colon keys', () => {
    for (const vector of catalog.identifier_encoding.vectors) {
      expect(encodeMetricAnalysisSubjectID(vector.raw)).toBe(vector.encoded)
      expect(decodeMetricAnalysisSubjectID(vector.encoded)).toBe(vector.raw)
      expect(vector.encoded).not.toContain('.')
      expect(vector.encoded).not.toContain(':')
    }
    expect(encodeMetricAnalysisSubjectID('domain:reasoning')).toBe('u-ZG9tYWluOnJlYXNvbmluZw')
    expect(encodeMetricAnalysisSubjectID('classifier:risk:RISKY')).toBe(
      'u-Y2xhc3NpZmllcjpyaXNrOlJJU0tZ',
    )
  })

  it('fails closed for unknown, malformed, retired, and ambiguous metric ids', () => {
    for (const metricID of [
      'routing.made_up_accuracy',
      'model_pool.arm.u-abc.quality',
      'capacity.level.0.success_rate',
      'routing_recipe.e2.feasible_oracle_recall_at_65',
      ...RETIRED_METRIC_ANALYSIS_IDS,
    ]) {
      expect(() => resolveMetricAnalysisCatalog(metricID)).toThrow()
    }

    const ambiguous = JSON.parse(METRIC_ANALYSIS_CATALOG_SOURCE) as CatalogFixture
    ambiguous.dynamic_families[1].literal_prefix = ambiguous.dynamic_families[0].literal_prefix
    ambiguous.dynamic_families[1].pattern = ambiguous.dynamic_families[0].pattern
    expect(() => validateMetricAnalysisCatalogSource(JSON.stringify(ambiguous))).toThrow(
      /prefixes overlap/,
    )
  })

  it('returns the exact estimator contract instead of inferring from the metric name', () => {
    expect(resolveMetricAnalysisCatalog('routing.accuracy').specification).toMatchObject({
      estimator_id: 'deterministic-routing-case-observed-ratio',
      estimator_version: 'v1',
      analysis_unit: 'route_case',
      cluster_unit: 'case',
      weighting: 'uniform_case',
      missingness: 'fail_closed',
      exclusion_policy: 'exclude_unavailable_evidence',
    })
  })
})
